sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/webapp/AadOAuth2UserServiceTest.java

```diff
@@ -0,0 +1,221 @@
+import org.springframework.security.oauth2.core.AuthenticationMethod;
+import org.springframework.security.oauth2.core.AuthorizationGrantType;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.oidc.IdTokenClaimNames;
+import org.springframework.security.oauth2.core.oidc.OidcIdToken;
+import org.springframework.security.oauth2.core.oidc.OidcScopes;
+import org.springframework.security.oauth2.core.oidc.StandardClaimNames;
+import org.springframework.security.oauth2.core.oidc.user.DefaultOidcUser;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import javax.servlet.http.HttpSession;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatIllegalArgumentException;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.RETURNS_DEEP_STUBS;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Tests for {@link AadOAuth2UserService}.
+ */
+class AadOAuth2UserServiceTest {
+    private ClientRegistration.Builder clientRegistrationBuilder;
+    private OidcIdToken idToken;
+    private AadOAuth2UserService aadOAuth2UserService;
+    private OAuth2AccessToken accessToken;
+    private Map<String, Object> idTokenClaims = new HashMap<>();
+    private GraphClient graphClient;
+    private AadAuthenticationProperties properties;
+
+    private static final String DEFAULT_OIDC_USER = "defaultOidcUser";
+
+
+    @BeforeEach
+    void setup() {
+
+        clientRegistrationBuilder = ClientRegistration
+            .withRegistrationId("registrationId")
+            .clientName("registrationId")
+            .authorizationGrantType(AuthorizationGrantType.AUTHORIZATION_CODE)
+            .redirectUri("redirectUri")
+            .userInfoUri(null)
+            .clientId("cliendId")
+            .clientSecret("clientSecret")
+            .userInfoAuthenticationMethod(AuthenticationMethod.HEADER)
+            .authorizationUri("authorizationUri")
+            .tokenUri("tokenUri");
+
+        this.accessToken = TestOAuth2AccessTokens.scopes(OidcScopes.OPENID, OidcScopes.PROFILE);
+
+        idTokenClaims.put(IdTokenClaimNames.ISS, "https://provider.com");
+        idTokenClaims.put(IdTokenClaimNames.SUB, "subject1");
+        idTokenClaims.put(StandardClaimNames.NAME, "user1");
+        idTokenClaims.put(StandardClaimNames.EMAIL, "user1@example.com");
+
+        this.idToken = new OidcIdToken("access-token", Instant.MIN, Instant.MAX, idTokenClaims);
```

Comment:
seems like this idToken isn't used by each test case, we can consider making this instantiated in each test method instead.

Reply:
I think loadUserWithCustomAuthorities() is the special one.
Other test cases could leverage the code in setup().

Reply:
If a variable is not fit for all cases, we should narrow the scope to method. It's okay to have some duplication in the UT. Which will make each test case easy to read.